1216 - Config File Docs (#1912)
defenseunicorns/zarf: docs/2-the-zarf-cli/2-declarative-cli-configuration.md

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import ExampleYAML from '@site/src/components/ExampleYAML';
import FetchFileCodeBlock from '@site/src/components/FetchFileCodeBlock';

> # Declarative CLI Configuration

Reviewer: ```suggestion
# Zarf Config Files
```
Not sure if this is the best wording (may be good to get more feedback from others) but I personally find declarative to be weird to use in a lot of places - config files do help with declaring configuration but they aren't the only way to do so - also most of our left sidebar titles are more concrete things (package components, zarf package schema) rather than philosophical concepts.

## Overview

> Users can configure the `zarf init`, `zarf package create`, and `zarf package deploy` command flags, as well as global flags (with the exception of `--confirm`), through a config file to help execute commands more declaratively.

Reviewer: ```suggestion
Users can use a config file to easily control flags for `zarf init`, `zarf package create`, and `zarf package deploy` commands, as well as global flags (excluding `--confirm`), enabling a more straightforward and declarative workflow.
```

> By default, Zarf searches for a config file named `zarf-config.toml` in the current working directory. You can generate a config template for use by Zarf by executing the command `zarf prepare generate-config`, with an optional filename, in any of the supported formats, including `toml`, `json`, `yaml`, `ini` and `props`. For instance, to create a template config file with the `my-cool-env` in the yaml format, you can use the command `zarf prepare generate-config my-cool-env.yaml`.

Reviewer: ```suggestion
Zarf supports config files written in common configuration file formats including `toml`, `json`, `yaml`, `ini` and `props`, and by default Zarf will look for a file called `zarf-config` with one of these filenames in the current working directory.  To generate a blank config file you can run `zarf prepare generate-config` with an optional output filename/format.  For example, to create an empty config file with the `my-cool-env` in the yaml format, you can use `zarf prepare generate-config my-cool-env.yaml`.
```

- by Zarf by is awkward wording
- executing the command `zarf ...` is over-explanatory
- filenames might be best explained first since Zarf will always look for all of them

> To use a custom config file, set the `ZARF_CONFIG` environment variable to the path of the desired config file. For example, to use the `my-cool-env.yaml` config file, you can set the `ZARF_CONFIG` environment variable to `my-cool-env.yaml`. The `ZARF_CONFIG` environment variable can be set either in the shell or in the `.env` file in the current working directory. Note that the `ZARF_CONFIG` environment variable takes precedence over the default config file.

Reviewer: ```suggestion
To use a custom config filename, set the `ZARF_CONFIG` environment variable to the config file's path. For example, to use the `my-cool-env.yaml` config file in the current working directory, you can set the `ZARF_CONFIG` environment variable to `my-cool-env.yaml`. The `ZARF_CONFIG` environment variable can be set either in the shell or in a `.env` file in the current working directory. Note that the `ZARF_CONFIG` environment variable takes precedence over the default config file path.
```

> Additionally, you can also set any supported config parameter via env variable using the `ZARF_` prefix. For instance, you can set the `zarf init` `--storage-class` flag via the env variable by setting the `ZARF_INIT.STORAGE_CLASS` environment variable. Note that the `ZARF_` environment variable takes precedence over the config file.

Reviewer: ```suggestion
Additionally, you can set any supported config parameter via an environment variable using the `ZARF_` prefix. For example, you can set the `zarf init` `--storage-class` flag by setting the `ZARF_INIT_STORAGE_CLASS` environment variable. Note that the `ZARF_` environment variable takes precedence over a config file.
```

While config files set default values, these values can still be overwritten by command line flags. For example, if the config file sets the log level to `info` and the command line flag is set to `debug`, the log level will be set to `debug`. The order of precedence for command line configuration is as follows:

1. Command line flags
2. Environment variables
3. Config file
4. Default values

For additional information, see the [Config File Example](../../examples/config-file/README.md).

## Config File Location

To use a custom config file, set the `ZARF_CONFIG` environment variable to the path of the desired config file. For example, to use the my-cool-env.yaml config file, you can set the `ZARF_CONFIG` environment variable to `my-cool-env.yaml`. The `ZARF_CONFIG` environment variable can be set either in the shell or in the .env file in the current working directory. Note that the ZARF_CONFIG environment variable takes precedence over the default config file.

> It will also pickup config files from either your current working directory or `~/.zarf/` if you don't specify a config file.

Reviewer: This mostly repeats the above and the extra information should probably just be combined there.

Author: I agree with you, i reduced the output since it was mostly repeated. LMK if the reduced content makes sense to you.

## Config File Examples

<Tabs queryString="init-file-examples">
<TabItem value="yaml">

> <ExampleYAML src={require('../../examples/config-file/zarf-config.yaml')} fileName="zarf-config.yaml" />

Reviewer: This should just use FetchFileCodeBlock
